From 36e93f4b5ba4c50f62639d53474c26069c138bb1 Mon Sep 17 00:00:00 2001
From: Martin Babinsky <mbabinsk@redhat.com>
Date: Thu, 23 Jul 2015 15:45:35 +0200
Subject: [PATCH] ACI plugin: correctly parse bind rules enclosed in
 parentheses

Since bind rule such as `(userdn = "ldap:///anyone")` is also a valid
statement, the ipalib ACI parser was updated to handle this case.

https://fedorahosted.org/freeipa/ticket/5037

Reviewed-By: Martin Basti <mbasti@redhat.com>
---
 ipalib/aci.py | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/ipalib/aci.py b/ipalib/aci.py
index 53c0053c66ec301e40be605f84c87ddc388450a2..8b0213ed0a2d24b4b5d1a8b42ede3c2d3d59c645 100755
--- a/ipalib/aci.py
+++ b/ipalib/aci.py
@@ -27,10 +27,11 @@ import ldap
 ACIPat = re.compile(r'\(version\s+3.0\s*;\s*acl\s+\"([^\"]*)\"\s*;\s*([^;]*);\s*\)', re.UNICODE)
 
 # Break the permissions/bind_rules out
-PermPat = re.compile(r'(\w+)\s*\((.*)\)\s+(.*)', re.UNICODE)
+PermPat = re.compile(r'(\w+)\s*\(([^()]*)\)\s*(.*)', re.UNICODE)
 
 # Break the bind rule out
-BindPat = re.compile(r'([a-zA-Z0-9;\.]+)\s*(\!?=)\s*(.*)', re.UNICODE)
+BindPat = re.compile(r'\(?([a-zA-Z0-9;\.]+)\s*(\!?=)\s*\"(.*)\"\)?',
+                     re.UNICODE)
 
 ACTIONS = ["allow", "deny"]
 
@@ -194,6 +195,9 @@ class ACI:
         self.target['target']['operator'] = operator
 
     def set_bindrule(self, bindrule):
+        if bindrule.startswith('(') != bindrule.endswith(')'):
+            raise SyntaxError("non-matching parentheses in bindrule")
+
         match = BindPat.match(bindrule)
         if not match or len(match.groups()) < 3:
             raise SyntaxError, "malformed bind rule"
-- 
2.6.4

